File: shared/main/java/net/ccbluex/liquidbounce/features/module/modules/combat/TpAura.java
package net.ccbluex.liquidbounce.features.module.modules.combat;

import static org.lwjgl.opengl.GL11.*;

import java.awt.*;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import net.ccbluex.liquidbounce.LiquidBounce;
import net.ccbluex.liquidbounce.api.enums.BlockType;
import net.ccbluex.liquidbounce.api.enums.EnumFacingType;
import net.ccbluex.liquidbounce.api.enums.MaterialType;
import net.ccbluex.liquidbounce.api.minecraft.block.state.IIBlockState;
import net.ccbluex.liquidbounce.api.minecraft.client.block.IBlock;
import net.ccbluex.liquidbounce.api.minecraft.client.entity.IEntity;
import net.ccbluex.liquidbounce.api.minecraft.client.entity.IEntityLivingBase;
import net.ccbluex.liquidbounce.api.minecraft.network.play.client.ICPacketPlayerDigging.WAction;
import net.ccbluex.liquidbounce.api.minecraft.network.play.client.ICPacketUseEntity;
import net.ccbluex.liquidbounce.api.minecraft.util.WBlockPos;
import net.ccbluex.liquidbounce.event.AttackEvent;
import net.ccbluex.liquidbounce.event.EventTarget;
import net.ccbluex.liquidbounce.event.Render3DEvent;
import net.ccbluex.liquidbounce.event.UpdateEvent;
import net.ccbluex.liquidbounce.features.module.Module;
import net.ccbluex.liquidbounce.features.module.ModuleCategory;
import net.ccbluex.liquidbounce.features.module.ModuleInfo;
import net.ccbluex.liquidbounce.utils.CPSCounter;
import net.ccbluex.liquidbounce.utils.CPSCounter.MouseButton;
import net.ccbluex.liquidbounce.utils.EntityUtils;
import net.ccbluex.liquidbounce.utils.ImmutableVec3;
import net.ccbluex.liquidbounce.utils.block.BlockUtils;
import net.ccbluex.liquidbounce.utils.extensions.PlayerExtensionKt;
import net.ccbluex.liquidbounce.utils.pathfinding.PathFinder;
import net.ccbluex.liquidbounce.utils.render.RenderUtils;
import net.ccbluex.liquidbounce.utils.timer.MSTimer;
import net.ccbluex.liquidbounce.utils.timer.TimeUtils;
import net.ccbluex.liquidbounce.value.BoolValue;
import net.ccbluex.liquidbounce.value.FloatValue;
import net.ccbluex.liquidbounce.value.IntegerValue;
import net.ccbluex.liquidbounce.value.ListValue;

/**
 * LiquidBounce Hacked Client A minecraft forge injection client using Mixin
 *
 * @author LeakedPvP
 * @game   Minecraft
 */
@ModuleInfo(name = "TpAura", description = "InfiniteAura from Sigma 4.1.", category = ModuleCategory.COMBAT)
public class TpAura extends Module
{
	final IntegerValue maxCPS = new IntegerValue("MaxCPS", 6, 1, 20)
	{
		@Override
		public void onChanged(final Integer prevValue, final Integer newValue)
		{
			final int i = minCPS.get();
			if (i > newValue)
				set(i);
			attackDelay = TimeUtils.randomClickDelay(i, get());
		}
	};
	final IntegerValue minCPS = new IntegerValue("MinCPS", 6, 1, 20)
	{
		@Override
		public void onChanged(final Integer prevValue, final Integer newValue)
		{
			final int i = maxCPS.get();
			if (i < newValue)
				set(i);
			attackDelay = TimeUtils.randomClickDelay(get(), i);
		}
	};
	private final BoolValue pathEspValue = new BoolValue("PathESP", true);
	private final IntegerValue pathEspTime = new IntegerValue("PathESPTime", 1000, 100, 3000);
	private final FloatValue rangeValue = new FloatValue("Range", 30.0F, 6.1F, 1000.0F);
	private final IntegerValue hurtTimeValue = new IntegerValue("HurtTime", 10, 0, 10);
	private final IntegerValue maxTargetsValue = new IntegerValue("MaxTargets", 4, 1, 50);
	private final IntegerValue maxDashDistanceValue = new IntegerValue("DashDistance", 5, 1, 10);
	private final ListValue autoBlockValue = new ListValue("AutoBlock", new String[]
	{
			"Off", "Fake", "Packet", "AfterTick"
	}, "Packet");
	private final BoolValue swingValue = new BoolValue("Swing", true);

	// Attack Delay
	private final MSTimer attackTimer = new MSTimer();
	private long attackDelay = TimeUtils.randomClickDelay(minCPS.get(), maxCPS.get());

	// Paths
	private final List<List<ImmutableVec3>> targetPaths = new ArrayList<>();

	// Targets
	public List<IEntityLivingBase> currentTargets = new CopyOnWriteArrayList<>();
	public IEntityLivingBase currentTarget;
	private List<ImmutableVec3> currentPath = new ArrayList<>();

	// Blocking Status
	public boolean clientSideBlockingStatus;
	private boolean serverSideBlockingStatus;

	@Override
	public final void onEnable()
	{
		currentTargets.clear();
	}

	@Override
	public final void onDisable()
	{
		currentTargets.clear();
		clientSideBlockingStatus = false;
	}

	@EventTarget
	public final void onUpdate(final UpdateEvent event)
	{
		if (mc.getThePlayer() == null)
			return;

		currentTargets = getTargets();

		if (attackTimer.hasTimePassed(attackDelay))
			if (!currentTargets.isEmpty())
			{
				targetPaths.clear();

				if (canBlock() && (mc.getThePlayer().isBlocking() || !"Off".equalsIgnoreCase(autoBlockValue.get())))
					clientSideBlockingStatus = true;

				for (int targetIndex = 0, targetCount = currentTargets.size() > maxTargetsValue.get() ? maxTargetsValue.get() : currentTargets.size(); targetIndex < targetCount; targetIndex++)
				{
					currentTarget = currentTargets.get(targetIndex);
					final ImmutableVec3 from = new ImmutableVec3(mc.getThePlayer().getPosX(), mc.getThePlayer().getPosY(), mc.getThePlayer().getPosZ());
					final ImmutableVec3 to = new ImmutableVec3(currentTarget.getPosX(), currentTarget.getPosY(), currentTarget.getPosZ());

					currentPath = computePath(from, to);
					targetPaths.add(currentPath); // Used for path esp

					// Unblock before attack
					if (mc.getThePlayer().isBlocking() || "Packet".equalsIgnoreCase(autoBlockValue.get()) || serverSideBlockingStatus)
					{
						mc.getNetHandler().addToSendQueue(classProvider.createCPacketPlayerDigging(WAction.RELEASE_USE_ITEM, WBlockPos.Companion.getORIGIN(), classProvider.getEnumFacing(EnumFacingType.DOWN)));
						serverSideBlockingStatus = false;
					}

					// Travel to the target
					for (final ImmutableVec3 path : currentPath)
						mc.getNetHandler().getNetworkManager().sendPacketWithoutEvent(classProvider.createCPacketPlayerPosition(path.getX(), path.getY(), path.getZ(), true));

					LiquidBounce.eventManager.callEvent(new AttackEvent(currentTarget));

					CPSCounter.registerClick(MouseButton.LEFT);

					if (swingValue.get())
						mc.getThePlayer().swingItem();

					// Attack target
					mc.getNetHandler().addToSendQueue(classProvider.createCPacketUseEntity(currentTarget, ICPacketUseEntity.WAction.ATTACK));

					// Block after attack
					if (canBlock() && !serverSideBlockingStatus && (mc.getThePlayer().isBlocking() || "Packet".equalsIgnoreCase(autoBlockValue.get())))
					{
						mc.getNetHandler().addToSendQueue(classProvider.createCPacketPlayerBlockPlacement(mc.getThePlayer().getInventory().getCurrentItemInHand()));
						serverSideBlockingStatus = true;
					}

					// Travel back to the original position
					Collections.reverse(currentPath);
					for (final ImmutableVec3 path : currentPath)
						mc.getNetHandler().getNetworkManager().sendPacketWithoutEvent(classProvider.createCPacketPlayerPosition(path.getX(), path.getY(), path.getZ(), true));
				}

				attackTimer.reset();
				attackDelay = TimeUtils.randomClickDelay(minCPS.get(), maxCPS.get());
			}
			else
				clientSideBlockingStatus = false;
	}

	@EventTarget
	public final void onRender3D(final Render3DEvent event)
	{
		if (!currentPath.isEmpty() && pathEspValue.get())
		{
			for (final List<ImmutableVec3> targetPath : targetPaths)
				try
				{
					for (final ImmutableVec3 pos : targetPath)
						if (pos != null)
							drawPath(pos);
				}
				catch (final Exception e)
				{
					// it seems sometime there is unknown interruption on these codes.
				}

			if (attackTimer.hasTimePassed(pathEspTime.get()))
			{
				targetPaths.clear();
				currentPath.clear();
			}
		}
	}

	private boolean canBlock()
	{
		return mc.getThePlayer() != null && mc.getThePlayer().getHeldItem() != null && classProvider.isItemSword(mc.getThePlayer().getHeldItem().getItem());
	}

	private List<ImmutableVec3> computePath(ImmutableVec3 from, final ImmutableVec3 to)
	{
		if (!canPassThrow(new WBlockPos(from.getX(), from.getY(), from.getZ())))
			from = from.addVector(0, 1, 0);

		final PathFinder pathfinder = new PathFinder(from, to);
		pathfinder.compute();

		int i = 0;
		ImmutableVec3 lastPath = null;
		ImmutableVec3 lastEndPath = null;
		final List<ImmutableVec3> path = new ArrayList<>();
		final List<ImmutableVec3> pathFinderPath = pathfinder.getPath();
		for (final ImmutableVec3 currentPathFinderPath : pathFinderPath)
		{
			if (i == 0 || i == pathFinderPath.size() - 1) // If the current path node is start or end node
			{
				if (lastPath != null)
					path.add(lastPath.addVector(0.5, 0, 0.5));
				path.add(currentPathFinderPath.addVector(0.5, 0, 0.5));
				lastEndPath = currentPathFinderPath;
			}
			else
			{
				boolean canContinueSearching = true;
				final float maxDashDistance = maxDashDistanceValue.get();
				if (currentPathFinderPath.squareDistanceTo(lastEndPath) > maxDashDistance * maxDashDistance)
					canContinueSearching = false;
				else
				{
					final double minX = Math.min(lastEndPath.getX(), currentPathFinderPath.getX());
					final double minY = Math.min(lastEndPath.getY(), currentPathFinderPath.getY());
					final double minZ = Math.min(lastEndPath.getZ(), currentPathFinderPath.getZ());
					final double maxX = Math.max(lastEndPath.getX(), currentPathFinderPath.getX());
					final double maxY = Math.max(lastEndPath.getY(), currentPathFinderPath.getY());
					final double maxZ = Math.max(lastEndPath.getZ(), currentPathFinderPath.getZ());
					cordsLoop:
					for (int x = (int) minX; x <= maxX; x++)
						for (int y = (int) minY; y <= maxY; y++)
							for (int z = (int) minZ; z <= maxZ; z++)
								if (!PathFinder.checkPositionValidity(x, y, z, false))
								{
									canContinueSearching = false;
									break cordsLoop;
								}
				}
				if (!canContinueSearching)
				{
					path.add(lastPath.addVector(0.5, 0, 0.5));
					lastEndPath = lastPath;
				}
			}
			lastPath = currentPathFinderPath;
			i++;
		}
		return path;
	}

	private List<IEntityLivingBase> getTargets()
	{
		return mc.getTheWorld().getLoadedEntityList().stream().filter(classProvider::isEntityLivingBase).map(IEntity::asEntityLivingBase).filter(entity -> PlayerExtensionKt.getDistanceToEntityBox(mc.getThePlayer(), entity) <= rangeValue.get() && EntityUtils.isEnemy(entity, false) && entity.getHurtTime() <= hurtTimeValue.get()).sorted((o1, o2) -> (int) (o1.getDistanceToEntity(mc.getThePlayer()) * 1000 - o2.getDistanceToEntity(mc.getThePlayer()) * 1000)).collect(Collectors.toList());
	}

	private boolean canPassThrow(final WBlockPos pos)
	{
		final IIBlockState state = BlockUtils.getState(new WBlockPos(pos.getX(), pos.getY(), pos.getZ()));
		final IBlock block = state.getBlock();
		return classProvider.getMaterialEnum(MaterialType.AIR).equals(block.getMaterial(state)) || classProvider.getMaterialEnum(MaterialType.PLANTS).equals(block.getMaterial(state)) || classProvider.getMaterialEnum(MaterialType.VINE).equals(block.getMaterial(state)) || classProvider.getBlockEnum(BlockType.LADDER).equals(block) || classProvider.getBlockEnum(BlockType.WATER).equals(block) || classProvider.getBlockEnum(BlockType.FLOWING_WATER).equals(block) || classProvider.getBlockEnum(BlockType.WALL_SIGN).equals(block) || classProvider.getBlockEnum(BlockType.STANDING_SIGN).equals(block);
	}

	public final void drawPath(final ImmutableVec3 vec)
	{
		final double x = vec.getX() - mc.getRenderManager().getRenderPosX();
		final double y = vec.getY() - mc.getRenderManager().getRenderPosY();
		final double z = vec.getZ() - mc.getRenderManager().getRenderPosZ();
		final double width = 0.3;
		final double height = mc.getThePlayer().getEyeHeight();

		// pre3D
		glPushMatrix();
		glEnable(GL_BLEND);
		glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
		glShadeModel(GL_SMOOTH);
		glDisable(GL_TEXTURE_2D);
		glEnable(GL_LINE_SMOOTH);
		glDisable(GL_DEPTH_TEST);
		glDisable(GL_LIGHTING);
		glDepthMask(false);
		glHint(GL_LINE_SMOOTH_HINT, GL_NICEST);

		glLoadIdentity();
		mc.getEntityRenderer().setupCameraTransform(mc.getTimer().getRenderPartialTicks(), 2);
		final Color[] colors =
		{
				Color.black, Color.white
		};
		for (int i = 0; i < 2; i++)
		{
			RenderUtils.glColor(colors[i]);
			glLineWidth(3 - i * 2);
			glBegin(GL_LINE_STRIP);
			glVertex3d(x - width, y, z - width);
			glVertex3d(x - width, y, z - width);
			glVertex3d(x - width, y + height, z - width);
			glVertex3d(x + width, y + height, z - width);
			glVertex3d(x + width, y, z - width);
			glVertex3d(x - width, y, z - width);
			glVertex3d(x - width, y, z + width);
			glEnd();
			glBegin(GL_LINE_STRIP);
			glVertex3d(x + width, y, z + width);
			glVertex3d(x + width, y + height, z + width);
			glVertex3d(x - width, y + height, z + width);
			glVertex3d(x - width, y, z + width);
			glVertex3d(x + width, y, z + width);
			glVertex3d(x + width, y, z - width);
			glEnd();
			glBegin(GL_LINE_STRIP);
			glVertex3d(x + width, y + height, z + width);
			glVertex3d(x + width, y + height, z - width);
			glEnd();
			glBegin(GL_LINE_STRIP);
			glVertex3d(x - width, y + height, z + width);
			glVertex3d(x - width, y + height, z - width);
			glEnd();
		}

		// post3D
		glDepthMask(true);
		glEnable(GL_DEPTH_TEST);
		glDisable(GL_LINE_SMOOTH);
		glEnable(GL_TEXTURE_2D);
		glDisable(GL_BLEND);
		glPopMatrix();
		glColor4f(1, 1, 1, 1);
	}

	public final boolean isTarget(final IEntityLivingBase entity)
	{
		return !currentTargets.isEmpty() && IntStream.range(0, currentTargets.size() > maxTargetsValue.get() ? maxTargetsValue.get() : currentTargets.size()).anyMatch(i -> currentTargets.get(i).isEntityEqual(entity));
	}

	@Override
	public final String getTag()
	{
		return String.valueOf(maxDashDistanceValue.get());
	}
}

File: shared/main/java/net/ccbluex/liquidbounce/features/module/modules/combat/TpAura.kt
package net.ccbluex.liquidbounce.features.module.modules.combat

import net.ccbluex.liquidbounce.LiquidBounce
import net.ccbluex.liquidbounce.api.enums.BlockType
import net.ccbluex.liquidbounce.api.enums.EnumFacingType
import net.ccbluex.liquidbounce.api.enums.MaterialType
import net.ccbluex.liquidbounce.api.minecraft.client.entity.IEntity
import net.ccbluex.liquidbounce.api.minecraft.client.entity.IEntityLivingBase
import net.ccbluex.liquidbounce.api.minecraft.client.entity.player.IEntityPlayer
import net.ccbluex.liquidbounce.api.minecraft.network.IPacket
import net.ccbluex.liquidbounce.api.minecraft.network.play.client.ICPacketPlayerDigging
import net.ccbluex.liquidbounce.api.minecraft.network.play.client.ICPacketUseEntity
import net.ccbluex.liquidbounce.api.minecraft.util.WBlockPos
import net.ccbluex.liquidbounce.api.minecraft.util.WBlockPos.Companion.ORIGIN
import net.ccbluex.liquidbounce.api.minecraft.util.WVec3
import net.ccbluex.liquidbounce.api.minecraft.world.IWorld
import net.ccbluex.liquidbounce.event.*
import net.ccbluex.liquidbounce.features.module.Module
import net.ccbluex.liquidbounce.features.module.ModuleCategory
import net.ccbluex.liquidbounce.features.module.ModuleInfo
import net.ccbluex.liquidbounce.utils.CPSCounter
import net.ccbluex.liquidbounce.utils.EntityUtils
import net.ccbluex.liquidbounce.utils.EntityUtils.isEnemy
import net.ccbluex.liquidbounce.utils.block.BlockUtils.getState
import net.ccbluex.liquidbounce.utils.extensions.getDistanceToEntityBox
import net.ccbluex.liquidbounce.utils.misc.StringUtils
import net.ccbluex.liquidbounce.utils.pathfinding.PathFinder
import net.ccbluex.liquidbounce.utils.render.ColorUtils
import net.ccbluex.liquidbounce.utils.render.RenderUtils
import net.ccbluex.liquidbounce.utils.timer.MSTimer
import net.ccbluex.liquidbounce.utils.timer.TimeUtils
import net.ccbluex.liquidbounce.value.*
import org.lwjgl.opengl.GL11.*
import java.util.concurrent.CopyOnWriteArrayList
import kotlin.math.max
import kotlin.math.min

/**
 * LiquidBounce Hacked Client A minecraft forge injection client using Mixin
 *
 * @author LeakedPvP
 * @game   Minecraft
 * // TODO: Maximum packets per ticks limit
 */
@ModuleInfo(name = "TpAura", description = "InfiniteAura from Sigma 4.1.", category = ModuleCategory.COMBAT)
class TpAura : Module()
{
	/**
	 * Options
	 */
	private val cpsValue: IntegerRangeValue = object : IntegerRangeValue("CPS", 3, 5, 1, 20, "MaxCPS" to "MinCPS")
	{
		override fun onMaxValueChanged(oldValue: Int, newValue: Int)
		{
			attackDelay = TimeUtils.randomClickDelay(getMin(), newValue)
		}

		override fun onMinValueChanged(oldValue: Int, newValue: Int)
		{
			attackDelay = TimeUtils.randomClickDelay(newValue, getMax())
		}
	}

	private val rangeValue = FloatValue("Range", 30.0f, 6.1f, 1000.0f)
	private val hurtTimeValue = IntegerValue("HurtTime", 10, 0, 10)
	val maxTargetsValue = IntegerValue("MaxTargets", 4, 1, 50)
	private val maxDashDistanceValue = IntegerValue("DashDistance", 5, 1, 10)
	private val autoBlockValue = ListValue("AutoBlock", arrayOf("Off", "Fake", "Packet", "AfterTick"), "Packet")
	private val swingValue = BoolValue("Swing", true)

	private val pathEspGroup = ValueGroup("PathESP")
	private val pathEspEnabledValue = BoolValue("Enabled", true, "PathESP")
	private val pathEspTagValue = BoolValue("Tag", false)
	private val pathEspLineWidthValue = FloatValue("LineWidth", 1f, 0.5f, 2f)
	private val pathEspTimeValue = IntegerValue("KeepLength", 1000, 100, 3000, "PathESPTime")

	private val pathEspColorValue = RGBAColorValue("Color", 255, 179, 72, 255, listOf("PathESP-Red", "PathESP-Green", "PathESP-Blue", "PathESP-Alpha"))
	private val pathEspColorRainbowGroup = ValueGroup("Rainbow")
	private val pathEspColorRainbowEnabledValue = BoolValue("Enabled", false, "PathESP-Rainbow")
	private val pathEspColorRainbowSpeedValue = IntegerValue("Speed", 10, 1, 10, "PathESP-RainbowSpeed")
	private val pathEspColorRainbowOffsetValue = IntegerValue("IndexOffset", 0, -100, 100, "PathESP-RainbowIndexOffset")
	private val pathEspColorRainbowSaturationValue = FloatValue("Saturation", 1.0f, 0.0f, 1.0f, "PathESP-RainbowHSB-Saturation")
	private val pathEspColorRainbowBrightnessValue = FloatValue("Brightness", 1.0f, 0.0f, 1.0f, "PathESP-RainbowHSB-Brightness")

	/**
	 * Variables
	 */

	// Attack Delay
	private val attackTimer = MSTimer()
	var attackDelay = cpsValue.getRandomClickDelay()

	// Paths
	private val targetPaths = mutableListOf<List<WVec3>>()

	// Targets
	private var currentTargets: MutableList<IEntityLivingBase> = CopyOnWriteArrayList()
	var currentTarget: IEntityLivingBase? = null
	private var currentPath = mutableListOf<WVec3>()

	// Blocking Status
	var clientSideBlockingStatus = false
	var serverSideBlockingStatus = false

	var debug: String? = null

	init
	{
		pathEspColorRainbowGroup.addAll(pathEspColorRainbowEnabledValue, pathEspColorRainbowSpeedValue, pathEspColorRainbowOffsetValue, pathEspColorRainbowSaturationValue, pathEspColorRainbowBrightnessValue)
		pathEspGroup.addAll(pathEspEnabledValue, pathEspTagValue, pathEspLineWidthValue, pathEspTimeValue, pathEspColorValue, pathEspColorRainbowGroup)
	}

	override fun onEnable()
	{
		currentTargets.clear()
	}

	override fun onDisable()
	{
		currentTargets.clear()

		if (serverSideBlockingStatus)
		{
			val provider = classProvider

			mc.netHandler.addToSendQueue(provider.createCPacketPlayerDigging(ICPacketPlayerDigging.WAction.RELEASE_USE_ITEM, ORIGIN, provider.getEnumFacing(EnumFacingType.DOWN)))
			serverSideBlockingStatus = false
		}

		clientSideBlockingStatus = false
	}

	@EventTarget
	fun onUpdate(@Suppress("UNUSED_PARAMETER") event: UpdateEvent?)
	{
		val theWorld = mc.theWorld ?: return
		val thePlayer = mc.thePlayer ?: return
		val netHandler = mc.netHandler
		val networkManager = netHandler.networkManager

		currentTargets = getTargets(theWorld, thePlayer)

		if (attackTimer.hasTimePassed(attackDelay))
		{
			val provider = classProvider

			if (currentTargets.isNotEmpty())
			{
				targetPaths.clear()
				if (canBlock(thePlayer) && (thePlayer.isBlocking || !autoBlockValue.get().equals("Off", ignoreCase = true))) clientSideBlockingStatus = true

				val from = WVec3(thePlayer.posX, thePlayer.posY, thePlayer.posZ)
				var targetIndex = 0

				val targetCount = if (currentTargets.size > maxTargetsValue.get()) maxTargetsValue.get() else currentTargets.size
				val single = targetCount == 1

				while (targetIndex < targetCount)
				{
					val currentTarget = currentTargets[targetIndex]

					this.currentTarget = currentTarget

					val to = WVec3(currentTarget.posX, currentTarget.posY, currentTarget.posZ)

					currentPath = computePath(theWorld, from, to)
					targetPaths.add(currentPath) // Used for path esp

					if (single) debug = "pathCount: ${currentPath.size}"

					// Unblock before attack
					if (thePlayer.isBlocking || autoBlockValue.get().equals("Packet", ignoreCase = true) || serverSideBlockingStatus)
					{
						netHandler.addToSendQueue(provider.createCPacketPlayerDigging(ICPacketPlayerDigging.WAction.RELEASE_USE_ITEM, ORIGIN, provider.getEnumFacing(EnumFacingType.DOWN)))
						serverSideBlockingStatus = false
					}

					// Travel to the target
					currentPath.forEach { path ->
						networkManager.sendPacketWithoutEvent(provider.createCPacketPlayerPosition(path.xCoord, path.yCoord, path.zCoord, true))
					}

					currentPath.reverse()

					LiquidBounce.eventManager.callEvent(AttackEvent(currentTarget, currentPath.firstOrNull() ?: from))

					CPSCounter.registerClick(CPSCounter.MouseButton.LEFT)

					if (swingValue.get()) thePlayer.swingItem()

					// Make AutoWeapon compatible
					var sendAttack = true
					val attackPacket: IPacket = provider.createCPacketUseEntity(currentTarget, ICPacketUseEntity.WAction.ATTACK)
					val autoWeapon = LiquidBounce.moduleManager[AutoWeapon::class.java] as AutoWeapon

					if (autoWeapon.state)
					{
						val packetEvent = PacketEvent(attackPacket)
						autoWeapon.onPacket(packetEvent)
						if (packetEvent.isCancelled) sendAttack = false
					}
					if (sendAttack) netHandler.addToSendQueue(attackPacket)

					// Block after attack
					if (canBlock(thePlayer) && !serverSideBlockingStatus && (thePlayer.isBlocking || autoBlockValue.get().equals("Packet", ignoreCase = true)))
					{
						netHandler.addToSendQueue(provider.createCPacketPlayerBlockPlacement(thePlayer.inventory.getCurrentItemInHand()))
						serverSideBlockingStatus = true
					}

					// Travel back to the original position
					currentPath.forEach { path ->
						networkManager.sendPacketWithoutEvent(provider.createCPacketPlayerPosition(path.xCoord, path.yCoord, path.zCoord, true))
					}
					targetIndex++
				}

				attackTimer.reset()
				attackDelay = cpsValue.getRandomClickDelay()
			}
			else
			{
				if (serverSideBlockingStatus)
				{
					netHandler.addToSendQueue(provider.createCPacketPlayerDigging(ICPacketPlayerDigging.WAction.RELEASE_USE_ITEM, ORIGIN, provider.getEnumFacing(EnumFacingType.DOWN)))
					serverSideBlockingStatus = false
				}

				clientSideBlockingStatus = false
				currentTarget = null
			}
		}
	}

	@EventTarget
	fun onRender3D(@Suppress("UNUSED_PARAMETER") event: Render3DEvent?)
	{
		val thePlayer = mc.thePlayer ?: return

		val renderManager = mc.renderManager
		val viewerPosX = renderManager.viewerPosX
		val viewerPosY = renderManager.viewerPosY
		val viewerPosZ = renderManager.viewerPosZ

		if (currentPath.isNotEmpty() && pathEspEnabledValue.get())
		{
			val rainbow = pathEspColorRainbowEnabledValue.get()
			val saturation = pathEspColorRainbowSaturationValue.get()
			val brightness = pathEspColorRainbowBrightnessValue.get()
			val pathEspAlpha = pathEspColorValue.getAlpha()
			val rainbowOffsetVal = 400000000L + 40000000L * pathEspColorRainbowOffsetValue.get()
			val rainbowSpeed = pathEspColorRainbowSpeedValue.get()
			val customColor = pathEspColorValue.get()

			val entityRenderer = mc.entityRenderer

			targetPaths.forEachIndexed { index, targetPath ->
				val color = if (rainbow) ColorUtils.rainbowRGB(pathEspAlpha, index * rainbowOffsetVal, rainbowSpeed, saturation, brightness) else customColor

				glPushMatrix()
				glDisable(GL_TEXTURE_2D)
				glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA)
				glEnable(GL_LINE_SMOOTH)
				glEnable(GL_BLEND)
				glDisable(GL_DEPTH_TEST)
				entityRenderer.disableLightmap()

				RenderUtils.glColor(color)
				glLineWidth(pathEspLineWidthValue.get())
				glBegin(GL_LINE_STRIP)

				targetPath.forEach { path ->
					glVertex3d(path.xCoord - viewerPosX, path.yCoord - viewerPosY, path.zCoord - viewerPosZ)
				}

				RenderUtils.resetColor()
				glEnd()

				glEnable(GL_DEPTH_TEST)
				glDisable(GL_LINE_SMOOTH)
				glDisable(GL_BLEND)
				glEnable(GL_TEXTURE_2D)
				glPopMatrix()

				if (pathEspTagValue.get()) targetPath.firstOrNull()?.let { path ->
					targetPath.lastOrNull()?.let { lastPath ->
						RenderUtils.renderNameTag("${StringUtils.DECIMALFORMAT_2.format(WVec3(lastPath.xCoord, lastPath.yCoord, lastPath.zCoord).distanceTo(WVec3(path.xCoord, path.yCoord, path.zCoord)))}m in ${targetPath.size} moves", path.xCoord + 0.5, path.yCoord + 0.2, path.zCoord + 0.5)
					}
				}
			}
		}

		if (attackTimer.hasTimePassed(pathEspTimeValue.get().toLong()))
		{
			targetPaths.clear()
			currentPath.clear()
		}
	}

	private fun computePath(theWorld: IWorld, from: WVec3, to: WVec3): MutableList<WVec3>
	{
		var fromPos = from

		if (!canPassThrough(theWorld, WBlockPos(fromPos.xCoord, fromPos.yCoord, fromPos.zCoord))) fromPos = fromPos.addVector(0.0, 1.0, 0.0)

		val pathfinder = PathFinder(fromPos, to)
		pathfinder.compute(theWorld)

		var lastPath: WVec3? = null
		var lastEndPath: WVec3? = null
		val path = mutableListOf<WVec3>()
		val pathFinderPath = pathfinder.path

		pathFinderPath.forEachIndexed { i, currentPathFinderPath ->
			if (i == 0 || i == pathFinderPath.size - 1)
			{
				// If the current path node is start or end node
				if (lastPath != null) path.add((lastPath ?: return@forEachIndexed).addVector(0.5, 0.0, 0.5))

				path.add(currentPathFinderPath.addVector(0.5, 0.0, 0.5))
				lastEndPath = currentPathFinderPath
			}
			else
			{
				var canContinueSearching = true
				val maxDashDistance = maxDashDistanceValue.get().toFloat()
				val lastEndPathChecked = lastEndPath ?: return@forEachIndexed

				if (currentPathFinderPath.squareDistanceTo(lastEndPathChecked) > maxDashDistance * maxDashDistance) canContinueSearching = false
				else
				{
					val minX = min(lastEndPathChecked.xCoord, currentPathFinderPath.xCoord)
					val minY = min(lastEndPathChecked.yCoord, currentPathFinderPath.yCoord)
					val minZ = min(lastEndPathChecked.zCoord, currentPathFinderPath.zCoord)
					val maxX = max(lastEndPathChecked.xCoord, currentPathFinderPath.xCoord)
					val maxY = max(lastEndPathChecked.yCoord, currentPathFinderPath.yCoord)
					val maxZ = max(lastEndPathChecked.zCoord, currentPathFinderPath.zCoord)
					var x = minX.toInt()
					cordsLoop@ while (x <= maxX)
					{
						var y = minY.toInt()
						while (y <= maxY)
						{
							var z = minZ.toInt()
							while (z <= maxZ)
							{
								if (!PathFinder.checkPositionValidity(theWorld, x, y, z, false))
								{
									canContinueSearching = false
									break@cordsLoop
								}
								z++
							}
							y++
						}
						x++
					}
				}

				if (!canContinueSearching)
				{
					path.add((lastPath ?: return@forEachIndexed).addVector(0.5, 0.0, 0.5))
					lastEndPath = lastPath
				}
			}
			lastPath = currentPathFinderPath
		}

		return path
	}

	private fun getTargets(theWorld: IWorld, thePlayer: IEntity): MutableList<IEntityLivingBase>
	{
		val range = rangeValue.get().toDouble()
		val hurtTime = hurtTimeValue.get()
		return EntityUtils.getEntitiesInRadius(theWorld, thePlayer, range).asSequence().filter { isEnemy(it, false) }.map(IEntity::asEntityLivingBase).filter { it.hurtTime <= hurtTime }.filter { thePlayer.getDistanceToEntityBox(it) <= range }.sortedBy { it.getDistanceToEntity(thePlayer) * 1000 }.toMutableList()
	}

	fun isTarget(entity: IEntity?): Boolean = currentTargets.isNotEmpty() && (0 until if (currentTargets.size > maxTargetsValue.get()) maxTargetsValue.get() else currentTargets.size).any { currentTargets[it] == entity }

	override val tag: String
		get() = "${maxDashDistanceValue.get()}"

	companion object
	{
		private fun canBlock(thePlayer: IEntityPlayer): Boolean = thePlayer.heldItem != null && classProvider.isItemSword(thePlayer.heldItem?.item)

		private fun canPassThrough(theWorld: IWorld, pos: WBlockPos): Boolean
		{
			val state = getState(theWorld, WBlockPos(pos.x, pos.y, pos.z))
			val block = state.block

			val provider = classProvider

			return provider.getMaterialEnum(MaterialType.AIR) == block.getMaterial(state) || provider.getMaterialEnum(MaterialType.PLANTS) == block.getMaterial(state) || provider.getMaterialEnum(MaterialType.VINE) == block.getMaterial(state) || provider.getBlockEnum(BlockType.LADDER) == block || provider.getBlockEnum(BlockType.WATER) == block || provider.getBlockEnum(BlockType.FLOWING_WATER) == block || provider.getBlockEnum(BlockType.WALL_SIGN) == block || provider.getBlockEnum(BlockType.STANDING_SIGN) == block
		}
	}
}
